### PR TITLE
Fixes bug so that isTopStrand now deals properly with unmapped reads

### PR DESCRIPTION
**Pull request by fleharty**
### Description

_Give your PR a **concise** yet **descriptive** title_
_Please explain the changes you made here._
_Explain the **motivation** for making this change. What existing problem does the pull request solve?_
_Mention any issues fixed, addressed or otherwise related to this pull request, including issue numbers or hard links for issues in other repos._
_You can delete these instructions once you have written your PR description._

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

#### Content
- [x] Added or modified tests to cover changes and any new functionality
- [x] Edited the README / documentation (if applicable)
- [ ] All tests passing on Travis

#### Review
- [ ] Final thumbs-up from reviewer
- [ ] Rebase, squash and reword as applicable

For more detailed guidelines, see https://github.com/broadinstitute/picard/wiki/Guidelines-for-pull-requests

### CODE REVIEW

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 1]
```diff
@@ -103,13 +103,20 @@ static boolean isTopStrand(final SAMRecord rec) {
 
         // If the read pair are aligned to different contigs we use
         // the reference index to determine relative 5' coordinate ordering.
-        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex())) {
+        // Both the read and its mate should not have their unmapped flag set to true.
+        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex()) && !rec.getReadUnmappedFlag() && !rec.getMateUnmappedFlag()) {
```
Comment (yfarjoun) [inline src/main/java/picard/sam/markduplicates/UmiUtil.java]: I feel terrible writting this, but `getReferenceIndex` can return a null....

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 2]
```diff
@@ -103,13 +103,20 @@ static boolean isTopStrand(final SAMRecord rec) {
...
             return rec.getFirstOfPairFlag() == (rec.getReferenceIndex() < rec.getMateReferenceIndex());
         }
 
-        final int read5PrimeStart = (rec.getReadNegativeStrandFlag()) ? rec.getUnclippedEnd() : rec.getUnclippedStart();
-        final int mate5PrimeStart = (rec.getMateNegativeStrandFlag()) ? SAMUtils.getMateUnclippedEnd(rec) : SAMUtils.getMateUnclippedStart(rec);
-        return rec.getFirstOfPairFlag() == (read5PrimeStart < mate5PrimeStart);
+        final int readUnclippedStart = rec.getReadUnmappedFlag() ? 0 : rec.getUnclippedStart();
```
Comment (yfarjoun) [inline src/main/java/picard/sam/markduplicates/UmiUtil.java]: I'm really nit-picking now...but reading the code for `getUnclippedStart` it seems that in extreme cases one could get zero or even negative values returned. I think that in this case the use of an Integer and `null`  are appropriate.

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 3]
```diff
@@ -103,13 +103,20 @@ static boolean isTopStrand(final SAMRecord rec) {
...
+        final int readUnclippedEnd = rec.getReadUnmappedFlag() ? 0 : rec.getUnclippedEnd();
```
Comment (yfarjoun) [inline src/main/java/picard/sam/markduplicates/UmiUtil.java]: why 0 here?

File: src/main/java/picard/sam/markduplicates/UmiUtil.java (hunk 2 above)
Comment (yfarjoun) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: Also, it isn't clear why yo are doing all these gymnastics. perhaps a comment with some diagrams of read pairs would help understand.

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 4]
```diff
@@ -133,7 +148,17 @@ static void setMolecularIdentifier(final SAMRecord rec, final String assignedUmi
...
+            ReadStrand strand = getStrand(rec);
+            switch (strand) {
+                case TOP:
+                    molecularIdentifier.append(TOP_STRAND_DUPLEX);
+                    break;
+                case BOTTOM:
+                    molecularIdentifier.append(BOTTOM_STRAND_DUPLEX);
+                    break;
+                default:
+                    break;
```
Comment (yfarjoun) [inline src/main/java/picard/sam/markduplicates/UmiUtil.java]: comment here as to what this means and what will happen to the read.

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 5]
```diff
@@ -146,4 +171,14 @@ static void setMolecularIdentifier(final SAMRecord rec, final String assignedUmi
     static int getUmiLength(final String umi) {
         return umi.length() - StringUtils.countMatches(umi, UmiUtil.DUPLEX_UMI_DELIMITER);
     }
+
+    /**
+     * An enum to hold the ordinality of a read
```
Comment (yfarjoun) [inline src/main/java/picard/sam/markduplicates/UmiUtil.java]: this isn't the ordinality....

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 6]
```diff
@@ -82,34 +82,49 @@ static String getTopStrandNormalizedUmi(final SAMRecord record, final String umi
...
+        if (rec.getReadUnmappedFlag() || rec.getMateUnmappedFlag()) {
+            return ReadStrand.UNKNOWN;
+        }
 
         // If the read pair are aligned to different contigs we use
         // the reference index to determine relative 5' coordinate ordering.
-        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex())) {
-            return rec.getFirstOfPairFlag() == (rec.getReferenceIndex() < rec.getMateReferenceIndex());
+        // Both the read and its mate should not have their unmapped flag set to true.
+        if (!rec.getReferenceIndex().equals(rec.getMateReferenceIndex()) && !rec.getReadUnmappedFlag() && !rec.getMateUnmappedFlag()) {
```
[the remark this answers is not in the document]
Comment (fleharty) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: Nice find, thanks!

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 7]
```diff
@@ -82,34 +82,49 @@ static String getTopStrandNormalizedUmi(final SAMRecord record, final String umi
...
+            if (rec.getFirstOfPairFlag() == (rec.getReferenceIndex() < rec.getMateReferenceIndex())) {
```
[the remark this answers is not in the document]
Comment (fleharty) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: It is mentioned above,

        // If the read pair are aligned to different contigs we use
        // the reference index to determine relative 5' coordinate ordering.

Did you want something else?

File: src/main/java/picard/sam/markduplicates/UmiUtil.java [hunk 8]
```diff
@@ -82,34 +82,49 @@ static String getTopStrandNormalizedUmi(final SAMRecord record, final String umi
...
+            } else {
+                return ReadStrand.BOTTOM;
+            }
         }
 
         final int read5PrimeStart = (rec.getReadNegativeStrandFlag()) ? rec.getUnclippedEnd() : rec.getUnclippedStart();
         final int mate5PrimeStart = (rec.getMateNegativeStrandFlag()) ? SAMUtils.getMateUnclippedEnd(rec) : SAMUtils.getMateUnclippedStart(rec);
-        return rec.getFirstOfPairFlag() == (read5PrimeStart < mate5PrimeStart);
+
+        if(rec.getFirstOfPairFlag() == (read5PrimeStart <= mate5PrimeStart)) {
```
[the remark this answers is not in the document]
Comment (fleharty) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: I'm not sure I understand what you are getting at here.  Could you elaborate?

[the remark this answers is not in the document]
Comment (yfarjoun) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: in the unlikely even that `read5PrimeStart== mate5PrimeStart` reads will be assigned top-strand solely by whether they are read1 or read2 (as in, read1 will be Top and read2 will be bottom. is that what you want?

[the remark this answers is not in the document]
Comment (fleharty) [reply src/main/java/picard/sam/markduplicates/UmiUtil.java]: Yes, this would be the appropriate thing to do.

**Pull request merged**